### ISSUE TITLE
Problem reading with datetime fields, 'PX_SdnToGregorian' is not defined

### ISSUE THREAD

**Issue by matthewswain** [mherrmann/pypxlib#4]
Hi I've come across an issue when using this module.

Using 

> Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:38:48) [MSC v.1900 32 bit (Intel)] on win32
> on Windows 8.1 x64

pypxlib installed with PIP.

When printing a row with a datetime I get:
`Row(EffectiveDate='<error>', RateCode=1, Rate=17.5)`

If I try to use the EffectiveDate field I get:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\mattsw\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pypxlib\__init__.py", line 230, in __getitem__
    field = self._table.fields[item]
KeyError: 0
>>> t[0].EffectiveDate
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\mattsw\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pypxlib\__init__.py", line 226, in __getattr__
    return self[item]
  File "C:\Users\mattsw\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pypxlib\__init__.py", line 232, in __getitem__
    return field.deserialize(pxval)
  File "C:\Users\mattsw\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pypxlib\__init__.py", line 128, in deserialize
    return self._deserialize(pxval.value)
  File "C:\Users\mattsw\AppData\Local\Programs\Python\Python35-32\lib\site-packages\pypxlib\__init__.py", line 155, in _deserialize
    PX_SdnToGregorian(
NameError: name 'PX_SdnToGregorian' is not defined
```

I've opened up both pxlib.dll and pxlib_x64.dll in dependency walker and can't see PX_SdnToGregorian, so wondered if this might be to do with versioning of the DLL.

**Comment by mherrmann** [comment]
Hi,

I'm sorry I can't reproduce this issue on my machine. If you are on Python 3, `PX_SdnToGregorian` is defined in `pypxlib/pxlib_ctypes/py3.py`. Maybe the `hasattr(_libs['px'], 'PX_SdnToGregorian')` check on line 1556 of that files fails on your system. Please investigate further and let me know what you find!

Cheers

**Comment by matthewswain** [comment]
Hi Michael,

You're right about the `hasattr` call returning false. I'm using 32-bit python so it's using the 32-bit DLL but not finding that function apparently..? When I get time I'll see if I can compile the DLL from source and try with that...

Cheers,
Matthew

**Comment by shoyu63** [comment]
Hello,
I have the same problem on Windows (64 bits).
I can't find the functions from gregor.c in the dll even when I compile it myself.
Best wishes.

Stéphane.

**Comment by mherrmann** [comment]
Hi Stéphane,

I see - I'm sorry. I currently don't have time to fix this but if you fix it and submit a pull request I'd be happy to incorporate it.

Best
Michael

**Comment by shoyu63** [comment]
Hello Michael,

I'm working on this problem and I'll send you a patch in may.

Best wishes,
Stéphane.

**Comment by mherrmann** [comment]
Thanks - looking forward :)
Am 28.04.2016 9:53 vorm. schrieb "Stéphane" notifications@github.com:

**Comment by andyrisko** [comment]
Hi,
The bug is in pxlib.dll: SdnToGregorian and GregorianToSdn are not  exported from gregor.c.
They are mising proper definition like:
PXLIB_API void PXLIB_CALL PX_SdnToGregorian(long int sdn, int *pYear, int *pMonth, int *pDay);
PXLIB_API long int PXLIB_CALL PX_GregorianToSdn(int year, int month, int day);

Cheers,
Andras

**Comment by prisockb** [comment]
I am running into the same issue. Has anyone developed a patch or fix yet?

**Comment by pierrie** [comment]
Hi 

I also got `<error>` instead of an value for a TimestampField.

This fixed my problem.
```

in the __init__.py file of pypxlib
 (  /usr/local/lib/python2.7/dist-packages/pypxlib   -- where mine is located)

```
fix like this :

line   2  : 
```

#from datetime import date, time      
to 
from datetime import *     (can be cleaned up more)


```
line 225 : comment out : 

`return TimeField._deserialize_ms(int(pxval_value.dval / 86400))
`
and add this :

```
		s = int(pxval_value.dval)
		seconds = int(s // 1000)
		SECONDS_PER_DAY = 60*60*24
		ordinal, seconds = divmod(seconds, SECONDS_PER_DAY)
		try:
			d = datetime.fromordinal(ordinal)
			return d + timedelta(seconds=seconds)
		except ValueError:
			return "error"

```
in the :  def _deserialize(cls, pxval_value):   in  the    class TimestampField(Field):


to look like this : 


```
class TimestampField(Field):
	@classmethod
	def _deserialize(cls, pxval_value):
		#return TimeField._deserialize_ms(int(pxval_value.dval / 86400)) #org
		s = int(pxval_value.dval)
		seconds = int(s // 1000)
		SECONDS_PER_DAY = 60*60*24
		ordinal, seconds = divmod(seconds, SECONDS_PER_DAY)
		try:
			d = datetime.fromordinal(ordinal)
			return d + timedelta(seconds=seconds)
		except ValueError:
			return "error"
	@classmethod
	def _serialize_to(cls, value, pxval_value):
		pxval_value.lval = TimeField._serialize(value) * 86400.0

```







hope this helps someone else.

**Comment by ghost** [comment]
Try using this code in lines 171+ in __init__.py

```
	def _serialize_to(self, value, pxval_value):
#		sdn = PX_GregorianToSdn(value.year, value.month, value.day)
#		pxval_value.lval = sdn - 1721425
		sdn = date.toordinal(value)
		pxval_value.lval = sdn
```

**Comment by pacoescriba** [comment]
Thanks, Pierrie!!! It worked perfectly!!!

**Comment by z010107** [comment]
I have the same issue. My fix is:


    class DateField(Field):
	    def _deserialize(self, pxval_value):
		    return date.fromordinal(int(pxval_value.lval))

	    def _serialize_to(self, value, pxval_value):
		    sdn = date.toordinal(value)
		    pxval_value.lval = sdn

**Comment by progressify** [comment]
Same problem for me

```
Traceback (most recent call last):
  File "C:/Users/Antonio/PycharmProjects/django-task-manager/task_manager/utils/paradox_helper.py", line 91, in <module>
    returndata = DBQuery2(tempFolderName)
  File "C:/Users/Antonio/PycharmProjects/django-task-manager/task_manager/utils/paradox_helper.py", line 79, in DBQuery2
    dictrow[headers[i]] = row[headers[i]]
  File "C:\Users\Antonio\PycharmProjects\django-task-manager\myvenv\lib\site-packages\pypxlib\__init__.py", line 262, in __getitem__
    return field.deserialize(pxval)
  File "C:\Users\Antonio\PycharmProjects\django-task-manager\myvenv\lib\site-packages\pypxlib\__init__.py", line 139, in deserialize
    return self._deserialize(pxval.value)
  File "C:\Users\Antonio\PycharmProjects\django-task-manager\myvenv\lib\site-packages\pypxlib\__init__.py", line 240, in _deserialize
    date = DateField._deserialize_days(days)
  File "C:\Users\Antonio\PycharmProjects\django-task-manager\myvenv\lib\site-packages\pypxlib\__init__.py", line 169, in _deserialize_days
    PX_SdnToGregorian(
NameError: name 'PX_SdnToGregorian' is not defined
```

![Screenshot_5](https://user-images.githubusercontent.com/11472632/65947356-64165f00-e438-11e9-9e5c-723feab1a265.png)

**Comment by mattbaker-digital** [comment]
In addition to @z010107's edit: https://github.com/mherrmann/pypxlib/issues/4#issuecomment-406185937, I had to change `_deserialize_days()` to the following:
``` python
def _deserialize_days(self, days):
	field_date = date.fromordinal(days)
	# if all values are 0, this is a blank date. return None
	if all([field_date.year == 0, field_date.month == 0, field_date.day == 0]):
		return None
	return field_date
```

**Comment by mherrmann** [comment]
I'm still open to merging a Pull Request that fixes this :-)

**Comment by mattbaker-digital** [comment]
@mherrmann WIll look into this now that I've got more understanding of how the library is structured.

**Comment by mherrmann** [comment]
Yay! :partying_face:

**Comment by mherrmann** [comment]
Thanks to @mattbaker-digital, this should now be fixed in release 2.4. Thanks!
